stage #2.5/module #4 REST Basics - Spring Web/#6.REST API Tools for Testing, Design, and Development/swagger_for_designing_and_documenting_rest_api.md

```diff
@@ -0,0 +1,201 @@
+
+## Swagger in Action
+
+To enable Swagger in a RESTful application you should add the following dependencies:
+
+> implementation group: 'io.springfox', name: 'springfox-swagger2', version: '3.0.0'
+implementation group: 'io.springfox', name: 'springfox-swagger-ui', version: '3.0.0'
+
+Add Swagger configuration in a separate configuration class:
+
+```Java
+@Configuration
+@EnableSwagger2
+public class SwaggerConfiguration {
+
+    @Bean
+    public Docket api(){
+        return new Docket(DocumentationType.SWAGGER_2)
+                .select()
+                .apis(RequestHandlerSelectors.basePackage("mjc.newsapplication.controller"))
+                .paths(PathSelectors.any())
+                .build().apiInfo(apiEndPointInfo());
+    }
+
+    public ApiInfo apiEndPointInfo(){
+        return new ApiInfoBuilder().title("Application Rest API")
+                .description("News Application API")
+                .contact(new Contact("MJC Supervisor", "mjc/finalexam", "mjc@epam.com"))
+                .license("Apache 2.0")
+                .licenseUrl("http://www.apache.org/licenses/LICENSE-2.0.html")
+                .version("0.0.1-SNAPSHOT")
+                .build();
+    }
+}
+```
+Here, You have added metadata information about the REST API such as API name, author, website, license and so on. You have also instructed Swagger to only generate documentation for the components present in the mjc.newsapplication.controller package.
+
+Since we have enabled Swagger, let us see the documentation of our API endpoints done by Swagger. This is rendered through Swagger UI in the following link:
+
+http://localhost:8080//swagger-ui.html#/news-controller
+
+![](./media/swagger_default_documentation.png "Swagger default documentation")
+
+Here, Swagger has put together the following information:
+
+- Document metadata (API name, license, website, contact and so on)
+-All REST endpoints with default information it can infer from code. Note that endpoint descriptions are method names
+
+These are the default information. If you want, you can provide a detailed description and information about the endpoints and operations.
+
+## Documenting Rest Controller
+Now Let's document the REST controller of your application explicitly. Swagger provides several annotations to add documentation metadata that it pulls while generating the documentation.
+
+For each of the REST endpoint and its associated operations, You can provide ApiOperation and their various responses with ApiResponses annotations.
+
+```Java
+@RestController
+@RequestMapping("/api/v1/news")
+@Api(produces = "application/json", value = "Operations for creating, updating, retrieving and deleting news in the application")
+public class NewsController {
+
+    @Autowired
+    private NewsRepository newsRepository;
+
+    @PostMapping
+    @ApiOperation(value = "Create a piece of news", response = ResponseEntity.class)
```

Comment:
May it would be better to use 'News' as the response type directly? There is no need to use ResponseEntity and specify HttpStatus.OK (used by default) in a controller annotated with `@RestController`, also, the API documentation generated with `response = ResponseEntity.class` will not be descriptive enough